### PR TITLE
[SPARK-40914][CONNECT] Mark `org.apache.spark.sql.connect` to be private

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR proposes to mark non-public API as package private. E.g. private[connect].

### Why are the changes needed?

This is to control our API surface and don't expose non-public API.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @cloud-fan @HyukjinKwon

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 1]
```diff
@@ -26,11 +26,14 @@ import org.apache.spark.sql.connect.planner.DataTypeProtoConverter
...
  * A collection of implicit conversions that create a DSL for constructing connect protos.
+ *
+ * All classes in connect/dsl are considered an internal API to Spark Connect and are subject to
+ * change between minor releases.
  */
 
 package object dsl {
 
-  object expressions { // scalastyle:ignore
+  private[connect] object expressions { // scalastyle:ignore
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:36]: If we document this as an internal package, I think it's fine to remove `private[connect]`, see also SPARK-16964 and SPARK-16813

Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala]: Makes sense at least for this dsl package. Removed `private[connect]` in this file.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cloud-fan** [comment]
Thinking about this more, I think only the client is the user-facing API of Spark Connect. We can mark the entire `connect` package as private, and remove all the API tagging (Unstable, Since, etc.).

**Comment by amaliujia** [comment]
@cloud-fan 

This also makes sense. Proto but server is not. 

Can you share an example of how to mark the entire package as `private`?

**Comment by cloud-fan** [comment]
For catalyst, what we did is to not generate API doc for it: https://github.com/apache/spark/blob/master/project/SparkBuild.scala#L1179

No public doc means everything in this package is private.

**Review by itholic — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
+1 for @cloud-fan 's opinion about preventing document generations.

**Comment by amaliujia** [comment]
Thanks everyone. I choose to apply the suggestion that mark the entire server module as private. The server implementation shouldn't be public API anyway. Only proto and DataFrames in clients are public API.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 2]
```diff
@@ -26,6 +26,9 @@ import org.apache.spark.sql.connect.planner.DataTypeProtoConverter
 
 /**
  * A collection of implicit conversions that create a DSL for constructing connect protos.
+ *
+ * All classes in connect/dsl are considered an internal API to Spark Connect and are subject to
+ * change between minor releases.
```
Comment (amaliujia) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:31]: I still keep this to match Catalyst's Dsl.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for update.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

The failed test is completely unrelated to this change.

**Pull request closed without merging**